pkg/virt-api/rest/subresource.go

```diff
@@ -158,21 +164,181 @@ func (app *SubresourceAPIApp) RestartVMRequestHandler(request *restful.Request,
+			return
+		}
+	}
+
+	changeRequests := []v1.VirtualMachineStateChangeRequest{}
+	for _, cr := range vm.Status.StateChangeRequests {
+		changeRequests = append(changeRequests, cr)
+	}
+
+	changeRequests = append(changeRequests,
+		v1.VirtualMachineStateChangeRequest{Action: v1.StopRequest, UID: &vmi.UID},
+		v1.VirtualMachineStateChangeRequest{Action: v1.StartRequest},
+	)
+
+	changeRequestBody, err := json.Marshal(changeRequests)
+	if err != nil {
+		response.WriteError(http.StatusInternalServerError, err)
+		return
+	}
+	bodyString := fmt.Sprintf("{\"status\": {\"stateChangeRequests\": %s}}", changeRequestBody)
+
+	log.Log.Object(vm).V(4).Infof("Patching VM: %s", bodyString)
+	_, err = app.VirtCli.VirtualMachine(namespace).Patch(vm.GetName(), types.MergePatchType, []byte(bodyString))
+	if err != nil {
+		response.WriteError(http.StatusInternalServerError, fmt.Errorf("%v: %s", err, bodyString))
+		return
+	}
+
+	response.WriteHeader(http.StatusAccepted)
+}
+
+func (app *SubresourceAPIApp) StartVMRequestHandler(request *restful.Request, response *restful.Response) {
+	name := request.PathParameter("name")
+	namespace := request.PathParameter("namespace")
+
+	vm, code, err := app.fetchVirtualMachine(name, namespace)
+	if err != nil {
+		response.WriteError(code, err)
+		return
+	}
+
+	bodyString := ""
+
+	runStrategy, err := vm.RunStrategy()
+	if err != nil {
+		response.WriteError(http.StatusInternalServerError, err)
+		return
+	}
+	// RunStrategyHalted         -> spec.running = true
+	// RunStrategyManual         -> send start request
+	// RunStrategyAlways         -> doesn't make sense
+	// RunStrategyRerunOnFailure -> doesn't make sense
+	switch runStrategy {
+	case v1.RunStrategyHalted:
+		// Patch can't delete an existing field in the VM, so honor
+		// whichever one is currently being used.
+		if vm.Spec.RunStrategy != nil {
+			bodyString = fmt.Sprintf("{\"spec\":{\"runStrategy\": \"%s\"}}", v1.RunStrategyAlways)
 		} else {
+			bodyString = fmt.Sprintf("{\"spec\":{\"running\": true}}")
+		}
+	case v1.RunStrategyManual, v1.RunStrategyRerunOnFailure:
+		changeRequests := []v1.VirtualMachineStateChangeRequest{}
+		for _, cr := range vm.Status.StateChangeRequests {
+			changeRequests = append(changeRequests, cr)
+		}
+
+		changeRequests = append(changeRequests,
+			v1.VirtualMachineStateChangeRequest{Action: v1.StartRequest},
+		)
+
+		changeRequestBody, err := json.Marshal(changeRequests)
+		if err != nil {
 			response.WriteError(http.StatusInternalServerError, err)
+			return
 		}
+		bodyString = fmt.Sprintf("{\"status\": {\"stateChangeRequests\": %s}}", changeRequestBody)
+	case v1.RunStrategyAlways:
+		response.WriteError(http.StatusBadRequest, fmt.Errorf("VM is already running"))
 		return
 	}
-	response.WriteHeader(http.StatusOK)
+
+	log.Log.Object(vm).V(4).Infof("Patching VM: %s", bodyString)
+	_, err = app.VirtCli.VirtualMachine(namespace).Patch(vm.GetName(), types.MergePatchType, []byte(bodyString))
```

Comment:
I can't tell with certainty because I'm not 100% confident with how the MergePatchType behaves with lists.

are these Patch commands overwriting the entire stateChangeRequest list, or are they adding to it?

We have an issue either way though. 

If it's overwriting the list, then calling 'virtctl stop ... && virtctl start ...' could result in the VMI never stopping because we write over the array list.

If we're not overwriting the list, then calling 'virtctl restart && virtctl restart && virtctl restart' would cause the VMI to be queued with 3 restarts in sequence... it would certainly be a surprise if I hit restart in the UI multiple times in a row for that to be additive. The expectation would be for restart requests to be idempotent.